### ISSUE TITLE
player should not be automatically removed from Challenges when leaving a Guild or Party

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica-android#906]
Currently on the Android app, when a player leaves a Guild, they are automatically removed from any challenges they have joined that are hosted in that Guild. This should not happen - they should remain as participants of the challenges, even for those hosted in private guilds.

The same applies when the player is removed from a Guild - that should not remove them from challenges.

The same applies to leaving or being removed from a Party.

_(FYI, this bug also used to occur on the website but has been fixed there in https://github.com/HabitRPG/habitica/pull/8517. I'm going to ask other mods to test to see if it happens on the iOS app and if so I'll log an issue in that repo too.)_

**Comment by vIiRuS** [comment]
I don‘t think this is actually caused by the app? The app does nothing to
remove the player from challenges once they leave a group, so that must be
done somewhere on the server side.

-- 
Phillip Thelen
Sent with Airmail

**Comment by Alys** [comment]
That's what I had assumed initially because it made sense, but when I tested leaving a guild on the website, I wasn't removed from challenges, but on the app I was. Maybe the app is using an API route and the website isn't! I'm closing this and doing more testing.

**Comment by Alys** [comment]
The problem is that the API's [Leave a group](https://habitica.com/apidoc/#api-Group-LeaveGroup) route takes an optional body parameter `keepChallenges`, which can be either `remain-in-challenges` (the required option to fix this issue) or `leave-challenges` (the default). The API's code is here: https://github.com/HabitRPG/habitica/blob/7fccf59f505c89618753a8a60958fcdfa7cd213f/website/server/controllers/api-v3/groups.js#L707-L728

The Android app (and if needed the iOS app) could be modified to specify `keepChallenges = remain-in-challenges`, but I think it would be better if the API were changed to make that the default, since it's by far the safest option. The website's default behaviour is already to keep challenges, so a user's expectation would be that that would be the default no matter what tool they used to leave a guild.

This would be a breaking change to the API since it would modify the behaviour of any tools that don't provide a value for `keepChallenges`, but again, the change would produce safer, more expected behaviour for the user so I think this change is justified. My guess would be that most third-party developers wouldn't mind much - it's possible that if they aren't already specifying a value for `keepChallenges` then they haven't considered the impact on the user of challenges being left automatically, and wouldn't object to their tool's default behaviour giving a better experience for the user.

@paglias @TheHollidayInn What do you think? I'd like to see the API's default changed. It would reduce "bug" reports. :) If there's no complaints in a few days, I'll close this issue again and open one in the main repo for the API change.

(FYI, there's also a related parameter in that API route called `keep` for whether to keep or remove challenge tasks if you do leave the challenges, but it's already set to the safe default of `keep-all`.)

**Comment by paglias** [comment]
I'm keeping a list of things we should change deprecate for API V4 but I'd prefer not to make breaking changes in API v3

**Comment by Alys** [comment]
Altariel (be317a32-c027-4032-bbd5-25ccb21eb912) reported in the Report a Bug guild that they were removed from Challenges unexpectedly and has tracked it down to accidentally leaving the Guild they were hosted in. They suggest that a confirmation pop-up could be added when leaving a Guild. That would be a workaround for this Challenge issue since it's not desired to adjust the default behaviour of API v3, and would also help anyone who accidentally leaves a private group, since it's impossible for them to put themselves back into the group.

_"As added information, now that I'm back on my laptop and can check the desktop site, I found that I was no longer in the Read Your Books guild, which is where the TBR Bingo and TBR #10 challenges are. If I somehow accidentally left the guild on Android due to lag or something, would that have kicked me from the challenges?_

_"If so, can I suggest a confirmation pop-up for the app, as I definitely did not intend to leave the guild, and in fact, still found it on my app's guild list, though not the guild list on the website."_

**Comment by saraolson** [comment]
@Alys so is the current solution for this particular issue to just add a confirmation dialogue when a user leaves a guild that has a challenge they're a part of that just confirms 'hey, leaving this will kick you out of any challenges this guild owns, cancel/ok'?

not sure the timeline on an apiv4

**Comment by Alys** [comment]
No, it would be much faster to just change the app's code to use the keep parameter. It's probably a one-line change. That would also be significantly better for the players and consistent with what's done on the website and iOS app.

**Comment by saraolson** [comment]
ah okay thank you for clarifying

**Comment by Alys** [comment]
I'm reopening this because it's happening again. 

When you leave a Guild you do see the options for keeping or leaving its Challenges, however even if you select the keep option, you are still removed from the Challenges.

I've tested this twice with my own account, and both times I was removed from challenges after selecting to keep them. We also have a bug report in RaB that I've tracked down to be due to this. I've copied it below for reference but it doesn't have any specific technical details in it.

Gandharv, The Non-Existent   @JainGandharv (5731ecf3-d34d-4f80-bcc0-2cf838edfcc6) [android]:

_"Hello!_
_Recently, I was removed from a lot of challenges without any notice. I have ruled out any accidental exits because I don't remember exiting the challenges, I wasn't shown a keep tasks/remove tasks dialog box and there are atleast 7 challenges that I had joined but am no longer a participant of. I know the challenges haven't ended because there are other participants in then and they end at a later date. Also, the tasks from those challenges are still there and I can delete them and they don't have any megaphone symbol next to them (Before, they did). And joining those challenges again creates new tasks. Another thing to note is that for some reason I am still part of 6 (seemingly random) challenges. I can provide a list of those challenges (from which I was removed) but that will take some time._
_I use the Android Beta App and it got an update the same day (April 10th). (Edit: It has another update available now! They've been releasing these updates within 3 days of each other)_"

**Comment by Mike-Antonacci** [comment]
I have a possible idea why this might be happening, but could use input from someone with more knowledge. `GuildDetailFragment.kt` has `leaveGroup` take a Boolean for whether or not to keep challenges, but the API isn't looking for a Boolean, it's looking for a String ("remain-in-challenges", "leave-challenges").

So, what might be happening is that since `leaveGroup` is passing a Boolean, the API doesn't get what it's expecting and so goes with the default value of "leave-challenges", and the default value of "keep-all" for tasks, explaining why @Alys was removed from challenges no matter which button she pressed. it also explains how for me, even if I select to leave the challenge, I keep the tasks, as that is the default value for the API.

Also, I can't find how `leaveGroup` does anything with the API. I've traced that in `GuildDetailFragment.kt`, it calls `leaveGroup`  from `GroupViewModel.kt` which in turn gets  `leaveGroup` from `SocialRepository.kt` but I hit a dead end at this point.
